src/doc-utils.js

```diff
@@ -1,38 +1,48 @@
 "use strict";
-function iterDoc(topDoc, func) {
-  const docs = [ topDoc ];
-  while (docs.length !== 0) {
-    const doc = docs.pop();
-    let res = undefined;
-
-    if (typeof doc === "string") {
-      const res = func("string", doc);
-
-      if (res) {
-        return res;
-      }
-    } else {
-      const res = func(doc.type, doc);
-
-      if (res) {
-        return res;
-      }
+function traverseDocUpDown(doc, onEnter, onExit) {
+  if (doc.type === "concat") {
+    onEnter(doc);
```

Comment:
instead of repeating it everywhere, you could put `onEnter(doc)` on the first line and `onExit(doc)` at the end.